# Week Assignment 03 — Baseline vs. Model

This notebook trains the Random Forest momentum classifier and compares it against a simple rule-based baseline on the out-of-time validation window defined in notebook 02.

In [1]:
import sys
sys.path.insert(0, '..')
import pandas as pd
from src.features import build_train_valid
from src.scoring_model import train_and_evaluate

raw = pd.read_csv('../data/sample_search_data.csv', parse_dates=['date'])
train_df, valid_df = build_train_valid(raw)
out = train_and_evaluate(train_df, valid_df)
out['results']['baseline_accuracy'], out['results']['model_accuracy']

(0.465625, 0.9375)

## Full comparison

In [2]:
import pandas as pd
r = out['results']
pd.DataFrame({
    'baseline': [r['baseline_accuracy'], r['baseline_macro_f1']],
    'model':    [r['model_accuracy'], r['model_macro_f1']],
}, index=['accuracy', 'macro_f1'])

,baseline,model
accuracy,0.465625,0.937500
macro_f1,0.211798,0.939349


## Per-class report (model)

In [3]:
pd.DataFrame(r['model_classification_report']).T

,precision,recall,f1-score,support
declining,0.962025,0.974359,0.968153,78.0000
growing,0.964286,0.870968,0.915254,93.0000
stable,0.910828,0.959732,0.934641,149.0000
accuracy,0.937500,0.937500,0.937500,0.9375
macro avg,0.945713,0.935019,0.939349,320.0000
weighted avg,0.938844,0.937500,0.937175,320.0000


## Takeaways
- The baseline (position-slope sign only) reaches ~47% accuracy, only slightly better than random three-way guessing (~33%) — position slope alone is a weak signal on its own.
- The Random Forest, using position, CTR-gap, and impression/click trend features together, reaches ~94% accuracy and ~0.94 macro-F1 on the held-out, strictly-later validation window — a large, meaningful lift over the baseline on the *same* split.
- Next: turn these predictions into a ranked, explainable action list (capstone notebook).